src/rgw/rgw_op.cc

```diff
@@ -901,6 +901,17 @@ void rgw_build_iam_environment(rgw::sal::Store* store,
     s->env.emplace("aws:UserAgent", i->second);
   }
 
+  i = m.find("QUERY_STRING"); {
+    std::size_t pos = (i->second).find("versionId");
+    if( pos != string::npos) {
+      std::size_t pos2 = (i->second).find("&", pos);
+      if( pos2 == string::npos) //end of the string reached
+        s->env.emplace("s3:VersionId", (i->second).substr(pos + 10));
```

Comment:
yes , it  is reading till end of url length or '&' char,

Reply:
for eg: consider below scenarios
 ["QUERY_STRING"] = "versionId=zWsV8M0tgFHupK6zTpYzJ4r7XV3CkGr", 
 ["QUERY_STRING"] = "versionId=zWsV8M0tgFHupK6zTpYzJ4r7XV3CkGr&partNumber=1"

in case of  ["QUERY_STRING"] = "versionId=**zWsV8M0tgFHupK6zTpYzJ4r7XV3CkGr**",  the if if( pos2 == string::npos) become true and we extract till end of the string

in case of  ["QUERY_STRING"] = "versionId=**zWsV8M0tgFHupK6zTpYzJ4r7XV3CkGr**&partNumber=1" the else part will extract till the '&'